### ISSUE TITLE
wrap_c no longer types input args

### ISSUE THREAD

**Issue by bjarthur** [ihnorton/Clang.jl#187]
here is my `test.h`:

```
int bar(int *data);
```

and the resulting `test.jl`:

```
function bar(data)
    ccall((:bar, test), Cint, (Ptr{Cint},), data)
end
```

while technically correct, old versions of `Clang` used to output:

```
function bar(data::Ptr{Cint})    ###  notice the input arg is typed
    ccall((:bar, test), Cint, (Ptr{Cint},), data)
end
```

could we please have this feature back?

also, isn't the current guideline to use `Ref` in this case, not `Ptr`?

thanks.

**Comment by yuyichao** [comment]
This is correct. You shouldn't pass in a pointer.

**Comment by bjarthur** [comment]
@yuyichao why did you close this?

**Comment by yuyichao** [comment]
Because the change is intentional and the type won't be added back.

**Comment by bjarthur** [comment]
why not specify the type as `Ref`?

**Comment by yuyichao** [comment]
The input type parameter is dangerous. https://github.com/ihnorton/Clang.jl/issues/152.

`Ref` doesn't really make a difference here.

**Comment by yuyichao** [comment]
And FWIW, NIDAQ.jl is exactly where this is used incorrectly. It should be fixed to correctly define `cconvert` and `unsafe_convert`.

**Comment by bjarthur** [comment]
i completely don't understand your explanation of the problem in #152 .  what is the GC root mechanism.  how how does typing the inputs encourage calling ccall in the wrong way?  what is an "adaptor method"?

it would've been nice if you had filed an issue with NIDAQ.jl about this when this change was made.

**Comment by yuyichao** [comment]
> what is the GC root mechanism

https://docs.julialang.org/en/latest/manual/calling-c-and-fortran-code/#Auto-conversion:-1

> how does typing the inputs encourage calling ccall in the wrong way

Because

> You shouldn't pass in a pointer.

There's no way for the runtime to figure out where's the pointer come from at the ccall site.

> what is an "adaptor method"?

https://docs.julialang.org/en/latest/manual/calling-c-and-fortran-code/#Auto-conversion:-1

> it would've been nice if you had filed an issue with NIDAQ.jl about this when this change was made.

Well, sure, the change doesn't break anything though.

**Comment by ihnorton** [comment]
@bjarthur just to add some context, "GC rooting" means telling the garbage collector that a Julia object is alive and needs to be preserved *during garbage collection*. If an object is not rooted then the garbage collector may decide that there are no references to the object and free it (leading to memory corruption if you then do something with the pointer in C). The issue with passing in a pointer is that `ccall` will not know that it refers to a Julia object underneath, whereas if you pass a Julia object then the system will insert the necessary calls to `unsafe_convert` *and make sure that the pointee object is rooted* during the `ccall`.

```
obj = myobject()

foo(x::Ptr{MyObject}) = ccall(..., ..., (Ptr{MyObject},) x) # <- ccall doesn't know that this is a Julia object, because you probably had to call `pointer` (which does not create a root)

foo2(x) = ccall(..., ..., (Ptr{MyObject},), x) # <- ccall knows that this is a Julia object, and unsafe_convert overload allows you to tell ccall how to safely get a pointer, which will be done automatically with the necessary rooting.
```

**Comment by bjarthur** [comment]
thanks.  i can't say i'm even close to understanding, but i've reworked NIDAQ.jl's introspection code, so it works now.

**Comment by bjarthur** [comment]
the julia [docs](https://docs.julialang.org/en/latest/manual/calling-c-and-fortran-code/#When-to-use-T,-Ptr{T}-and-Ref{T}-1) are pretty clear that `Ref` should be generally used C functions accepting pointers as it has the flexibility for the memory to be managed by either C or julia, whereas `Ptr` is strictly only for C memory management.

given this, in my example above, shouldn't the `ccall` generated by Clang specify a `Ref` as the input arg, not a `Ptr`?

**Comment by yuyichao** [comment]
Doesn't matter for custom types.

**Comment by bjarthur** [comment]
the type in the example `int`.  what do you mean by "custom types"?

**Comment by yuyichao** [comment]
Still doesn't matter unless it encourage you to pass in a Ptr
